### ISSUE TITLE
Being able to use the same AlgaSynthDefs for both AlgaNode and AlgaPattern

### ISSUE THREAD

**Issue by thresholdpeople** [vitreo12/AlgaLib#64]
Moving our discussion from PMs on scsynth.org to here to track this better!

tldr- I've got a proposed solution that involves `Node`'s `release` method that I discuss at the very bottom, but first I just describe all of the things that I tried to make sure that I'm not missing anything and that my solution, or some other alternative will truly be necessary to implement something like this.
 
## The Issue:
Currently `AlgaNode` only supports defs which do not have any sort of self-freeing mechanism.

```
a = AlgaNode({ SinOsc.ar * EnvPerc.kr });

posts the following error: 

ERROR: AlgaNode: trying to instantiate the AlgaSynthDef 'alga_1013' which can free its synth. This is not supported for AlgaNodes, but it is for AlgaPatterns.
```

`AlgaPattern` is the opposite, it rejects any def which does not have a self-freeing mechanism:

```
a = AlgaPattern((def: { SinOsc.ar }));

posts:

ERROR: AlgaPattern: AlgaSynthDef 'alga_1015' can't free itself: it doesn't implement any Done action.
```

This means that effectively the same `AlgaSynthDef` cannot be repurposed. 

### AlgaDetectSilence:
From our discussion in PMs, one workaround that we discussed is to add `AlgaDetectSilence` to all defs, and then feed an envelope to the `amp` NamedControl parameter (which will be required to be included, though of course can have a different name) via `AlgaTemp`.

```
(
y = AP((
	def: { 
		var sig = SinOsc.ar(\freq.kr(440)) * \amp.kr(0.2);
		AlgaDetectSilence.ar(sig);
		sig;
	},
	amp: AlgaTemp({ EnvPerc.kr(\atk.kr(0.01), \rel.kr(0.25), \curve.kr(-2)) }),
	dur: 0.5,
)).play
)
```

The works: AlgaTemp creates a node which modulates the `amp` key, lowering the volume of the main def, and then allowing `AlgaDetectSilence` to kick in and free the node. 

I just tried using the same def for `AlgaNode`, and it threw the same error as mentioned at the start of this message. So this won't really work. 

### `doneAction` Argument
Another thing I tried is adding an envelope with a `NamedControl` for its `doneAction` argument. `{ SinOsc.ar * Env.perc(0.01, 1).kr(doneAction: \da.kr(0)) }`. This unfortunately doesn't work:

```
a = AN({ SinOsc.ar * Env.perc(0.01, 1).kr(doneAction: \da.kr(0)) })
returns: 
ERROR: AlgaNode: trying to instantiate the AlgaSynthDef 'alga_1060' which can free its synth. This is not supported for AlgaNodes, but it is for AlgaPatterns.
```
Surprisingly, everything is fine with `AlgaPattern`, *even if \da is left at its default 0 setting!*

```
a = AP((def: { SinOsc.ar * Env.perc(0.01, 1).kr(doneAction: \da.kr(0)) })).play // works!

a = AP((def: { SinOsc.ar * Env.perc(0.01, 1).kr(doneAction: \da.kr(0)) }, da: 0)).play // and this also works!
```
The plot thickens a little if removing the `NamedControl` and sticking a constant value in its place:
```
a = AN({ SinOsc.ar * Env.perc(0.01, 1).kr(doneAction: 0) }).play // this works!
a = AP((def: { SinOsc.ar * Env.perc(0.01, 1).kr(doneAction: 0) })).play // posts error
```
Thinking that perhaps `\da` somehow overlapped inside of an Alga class definition, I tried other symbols, but the results were the same as above. 

So it seems that if there is a variable in `doneAction`, Alga commandeers it and always sticks a freeing value in its place. 

## Sustained Envelopes (don't work with `AlgaPattern`):
One other thing that I noticed is that sustained envelopes don't work with `AlgaPattern`. I'm including this in our discussion here because I think that my idea, if possible, can allow for both AlgaSynthDef cross-compatibility between `AlgaNode` and `AlgaPattern` *and* allow for releasing sustaining envelopes.

Quoting Francesco from a PM - 

> Maybe, like standard Pbinds there could be a \sustain key that specifies when to send the release trigger to the synth. However, I really don’t want to have reserved keywords that the user might use in the AlgaSynthDef.

I totally agree with this. Especially in Patterns `\sustain` gets strange as depending on whether its used in an `adsr` or similar (sustains until `EnvGen`'s `gate` arg set to 0), or `linen` or similar (sustain here is a time value).. 

## Proposed Fix:
I'm assuming Alga is able to keep track of all nodes which it creates, and knows what is where. If it doesn't then my fix won't work without implementing that.

What if `AlgaPattern` had an additional argument, called `release` or something of the sort. When `release` receives a value, after that value of time elapses, the node receives a `release` message. Perhaps `release` can be set to either floats or an array. A float would of course be the hold duration, but in the case of an array: [holdDur, releaseTime].

`release` could default to `nil`, meaning that any def provided which already has self-freeing abilities works as expected. If `release` is provided a value and the def has a self-freeing mechanism, then it's like the case of if a SynthDef has two envelopes with `Done.freeSelf`: whichever happens first. 

I like this solution because as opposed to preventing Alga from commandeering a variable provided to `doneAction` it would prevent supplying `AlgaSynthDef`s with "dummy" self-freeing UGens. 

I also think that if this is possible, calling `release` on the node a synth is running on is as good as setting `gate = 0` to free a synth while keeping the node alive to accommodate for the release time.

What do you think?

**Comment by vitreo12** [comment]
Thank you for the very thourough analysis of the issue!

Let me unpack here my thoughts and an additional proposal:

First of all, I definitely agree that there should  be a consistent way to make `AlgaSynthDef`s work for both `AlgaNode`s and `AlgaPattern`s without having to redefine them to accomodate the presence / absence of a `Done.freeSelf` mechanism. 

>Thinking that perhaps \da somehow overlapped inside of an Alga class definition, I tried other symbols, but the results were the same as above.

>So it seems that if there is a variable in doneAction, Alga commandeers it and always sticks a freeing value in its place.

The problem with this is that Alga can't reliably know the value of `\da` when it creates the `AlgaSynthDef` (it's not a constant). Though, it assumes that the `AlgaSynthDef` will be able to free itself: it does not rely on the default value of 0.

> I'm assuming Alga is able to keep track of all nodes which it creates, and knows what is where. If it doesn't then my fix won't work without implementing that.

Yes, it surely does.

>What if AlgaPattern had an additional argument, called release or something of the sort. When release receives a value, after that value of time elapses, the node receives a release message. Perhaps release can be set to either floats or an array. A float would of course be the hold duration, but in the case of an array: [holdDur, releaseTime].

>release could default to nil, meaning that any def provided which already has self-freeing abilities works as expected. If release is provided a value and the def has a self-freeing mechanism, then it's like the case of if a SynthDef has two envelopes with Done.freeSelf: whichever happens first.

>I like this solution because as opposed to preventing Alga from commandeering a variable provided to doneAction it would prevent supplying AlgaSynthDefs with "dummy" self-freeing UGens.

>I also think that if this is possible, calling release on the node a synth is running on is as good as setting gate = 0 to free a synth while keeping the node alive to accommodate for the release time.

>What do you think?

So, I kind of do not like the solution of having that additional argument provided as is. I had another solution in mind, let me know what you think.

If the user writes an `AlgaSynthDef` that does not specify any `Done` mechanism, `AlgaSynthDef` could create an additional definition, say `\dummy_pattern`, which would be the one used whenever calling `\dummy` in an `AlgaPattern`. This additional definition would define an `\amp` parameter (if the user has not declared it in the code already) which can be tested against with `AlgaDetectSilence` to see when it's done. This way, the control of the length of the `Synth` is not handled when instantiating / replacing an `AlgaPattern`, but directly in the parameters.

Of course, this would also apply if using a `Function` as `def`:

```
(
y = AP((
	def: { SinOsc.ar(\freq.kr(440)) \amp.kr(1) },
	amp: AlgaTemp({ EnvPerc.kr(\atk.kr(0.01), \rel.kr(0.25), \curve.kr(-2)) }),
	dur: 0.5,
)).play
)
```

In this example, the `AlgaTemp` would control the `\amp` parameter (declared by the user in this case). When this is detected to be 0 by `AlgaDetectSilence` (declared by Alga in the background), the `Synth` can be freed.

This way, it also allows to dynamically change envelope shapes, sequence them, etc...

**Comment by vitreo12** [comment]
>I totally agree with this. Especially in Patterns \sustain gets strange as depending on whether its used in an adsr or similar (sustains until EnvGen's gate arg set to 0), or linen or similar (sustain here is a time value)..

Regarding sustains, I am at a loss for a solution to be honest. Perhaps, it can be a parameter that can be used only if the AlgaSynthDef explicitly has control over its own release?

**Comment by thresholdpeople** [comment]
> If the user writes an AlgaSynthDef that does not specify any Done mechanism, AlgaSynthDef could create an additional definition, say \dummy_pattern, which would be the one used whenever calling \dummy in an AlgaPattern (Note, the user would still use \dummy, he/she doesn't have to know about this additional definition declared in the background). This additional definition would define an \amp parameter (if the user has not declared it in the code already) which can be tested against with AlgaDetectSilence to see when it's done. This way, the control of the length of the Synth is not handled when instantiating / replacing an AlgaPattern, but directly in the parameters.

I like this because then it will allow you to very modular with the type of envelope you can use, AR, ADSR, etc. So I think this extends the functionality a lot. 

My one concern, and where I was thinking an arg might work better is that not only is it more typing, but also somewhat invisible (I mean that you technically don't *need* to use the \amp key in the Event, and there's nothing to remind you to do so). As opposed to an arg that's always in `AlgaPattern`, pops up in the arg hints, arg name, and helpfile. And it can be optional to set, just like most of the other args. 

The other benefit is that it does allow you to use sustaining envelopes with `AlgaPattern`. 

That being said, feeding in `AlgaTemp` also lets you do so, so I think your way also solves for both.

```
(
y = AP(( // EnvPerc here just to let AP run
	def: { Env.linen(0, 4, 4).kr(Done.freeSelf); SinOsc.ar(\freq.kr(440)) * \amp.kr(1).poll; },
	amp: AlgaTemp({ Env.adsr(\atk.kr(0.2), \dec.kr(0.3), \sus.kr(0.2), \rel.kr(0.75)).kr }),
	dur: 1,
)).play
)
```

**Comment by thresholdpeople** [comment]
One other thought on sustaining envs: and I'm not sure how difficult this is, and so it may not be worth it, and I don't love this idea, but maybe it's ok:

If `AlgaPattern` notices a sustained envelope in its def, perhaps it can add the Pbind-esque \sustain and gate release functionality then.

Or maybe just having `\sustain` as a reserved key like `\out` and `\gate` is fine? I feel like many folks who have used `Pbind` in the past should sort of expect it? If this is the case, I think that unlike `\out` and `\gate`, it should *not* be illegal to use it in your def (personally I think \gate should be allowed as well, but that's not a serious concern for me).

**Comment by vitreo12** [comment]
>If AlgaPattern notices a sustained envelope in its def, perhaps it can add the Pbind-esque \sustain and gate release functionality then.

I was thinking of this too. Also, because it's the only way that would work with the new mechanism proposed above. \sustain, then, would only work if you provided the Env in the definition, and not via \amp (AlgaTemp, or such)

**Comment by thresholdpeople** [comment]
> \sustain, then, would only work if you provided the Env in the definition, and not via \amp (AlgaTemp, or such), as that would be impossible to track / set reliably.

What if instead of killing off the `AlgaTemp` node a `release` method were triggered instead?

**Comment by vitreo12** [comment]
>What if instead of killing off the AlgaTemp node along with each Event instance, a release method were triggered instead?

When would trigger that though? With this proposal the AlgaTemp would "die" when it reaches 0, how can we track if it has sustain?

**Comment by vitreo12** [comment]
With \amp, the sustain would come from the AlgaTemp: how to send \sustain to the AlgaTemp? Maybe it could be a reserved keyword at all times and be sent to all active AlgaTemps too?

**Comment by thresholdpeople** [comment]
Ah yes, of course. The `AlgaPattern` node never gets freed since \amp never gets to 0, and thus `AlgaTemp` never gets the ability to get released, etc. 

Then, yes I agree, I think it would require to send sustain to `AlgaTemp`s as well.

**Comment by thresholdpeople** [comment]
What about using \legato instead of \sustain. I think there would be less chances of name clashes, and I think technically the way it works in Pbind is the end of legato triggers the release stage.

https://doc.sccode.org/Tutorials/A-Practical-Guide/PG_07_Value_Conversions.html
> legato
A fraction of the event's duration for which the synth should sustain. 1.0 means this synth will release exactly at the onset of the next; 0.5 means the last half of the duration will be a rest. Values greater than 1.0 produce overlapping notes. sustain = dur * legato * stretch.

**Comment by vitreo12** [comment]
Fantastic, I'll open a PR and work on these things. I'll get back here with more stuff once I get it working :)

**Comment by vitreo12** [comment]
So, I have implemented both things if you want to test them out on the `algasynthdef-patterns` branch which the PR is based on. Additionally to what we discussed on, I have also added an additional argument to `AlgaTemp`: `sampleAccurate`. This allows to trigger things like envelope with sample accuracy, otherwise they would be shifted back if the `def` was `sampleAccurate` but the `AlgaTemp` controlling the `amp` not.

The two test files are:

1. `TestAlgaSynthDef_Patterns.scd`:
```
//1
(
Alga.boot({
	AlgaSynthDef(\test, {
		SinOsc.ar(\freq.kr(440))
	}, sampleAccurate: true).add;

	s.sync;

	//a = AN(\test).play;

	b = AP((
		def: \test,
		amp: AlgaTemp({
			EnvPerc.ar(\atk.kr(0.01), \rel.kr(0.25), \curve.kr(-2), 0)
		}, sampleAccurate: true),
		dur: 0.5,
	)).play(chans:2)
});
)

//2
(
Alga.boot({
	y = AP((
		def: { SinOsc.ar(\freq.kr(440)) * \amp.kr(1) },
		amp: AlgaTemp({
			EnvPerc.kr(\atk.kr(0.01), \rel.kr(0.25), \curve.kr(-2), 0)
		}),
		dur: 0.5,
	)).play(chans:2)
})
)

//3
(
Alga.boot({
	y = AP((
		def: { SinOsc.ar(\freq.kr(440)) },
		amp: Pseq([
			AlgaTemp({
				EnvPerc.ar(\atk.kr(0.01), \rel.kr(0.25), \curve.kr(-2), 0) * 0.25
			}, sampleAccurate: true),
			AlgaTemp({
				EnvPerc.ar(\atk.kr(0.001), \rel.kr(0.1), \curve.kr(-2), 0) * 0.25
			}, sampleAccurate: true),
		], inf),
		freq: Pwhite(440, 880),
		dur: Pwhite(0.01, 0.2),
	)).play(chans:2)
})
)
```

2. `TestSustain.scd`:

```
//1
(
Alga.boot({
	a = AP((
		def: { SinOsc.ar * EnvGen.ar(Env.adsr, \gate.kr) }, //use can use \gate
		sustain: Pseq([1, 2], inf), //reserved keyword
		dur: 3
	)).play
})
)

//2
(
Alga.boot({
	a = AP((
		def: { SinOsc.ar },
		amp: AlgaTemp({ EnvGen.ar(Env.adsr, \gate.kr) }, sampleAccurate: true),
		sustain: Pseq([1, 2], inf),
		dur: 3
	)).play
})
)
```

**Comment by thresholdpeople** [comment]
You work very very quickly! I'm super excited to try this out. Heading to work shortly, but I'll play with it this evening. 

Thank you Francesco!

**Comment by thresholdpeople** [comment]
This is great! All the tests work for me. I'll have a go at running some of the things I've been making now. 

About sustain: what do you think about configuring it so that it's based on beats instead of seconds, for example:

```
~clock = TempoClock(1).permanent_(true);
Alga.boot({
	a = AP(( 
		def: { SinOsc.ar },
		amp: AlgaTemp({ EnvGen.ar(Env.adsr, \gate.kr) }, sampleAccurate: true),
		sustain: Pseq([1, 2], inf),
		dur: 3
	)).play
}, clock: ~clock);
)

~clock.tempo = 6
```
I can see it making sense either way.

**Comment by thresholdpeople** [comment]
One other thought about sustain-- what about (also) being able to derive it from `dur * stretch`? That way it doesn't have to be called explicitly. Might make sense to implement legato in this case as well? `sustain = dur * stretch * legato`.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by vitreo12** [comment]
> This is great! All the tests work for me. I'll have a go at running some of the things I've been making now.

Let me know how those go! 

> About sustain: what do you think about configuring it so that it's based on beats instead of seconds, for example:

Good point: i'll add an option in the form of a variable like `scheduleSustainToBeats`.

> One other thought about sustain-- what about (also) being able to derive it from `dur * stretch`? That way it doesn't have to be called explicitly (this would of course force it to be beat/tempo based rather than time). Might make sense to implement legato in this case as well? `sustain = dur * stretch * legato`.

This also would be nice. I'll look into it :)

**Comment by vitreo12** [comment]
Going back to this

>About sustain: what do you think about configuring it so that it's based on beats instead of seconds, for example:

I noticed that actually it already works as proposed, as the default clock used by Alga is TempoClock. One might want to use a SystemClock if needing to schedule things in seconds?

```
(
Alga.boot({
	a = AP(( 
		def: { SinOsc.ar },
		amp: AlgaTemp({ EnvGen.ar(Env.adsr, \gate.kr) }, sampleAccurate: true),
		sustain: Pseq([1, 2], inf),
		dur: 3
	)).play
});
)

a.clock.tempo=6
```

Or maybe I did not get fully what you meant

**Comment by vitreo12** [comment]
Ok, what I did whas to add a `schedSustainInSeconds` flag which allows to schedule sustain in seconds, instead that according to the clock of the `AlgaPattern`:

```
(
Alga.boot({
	a = AP(( 
		def: { SinOsc.ar },
		amp: AlgaTemp({ EnvGen.ar(Env.adsr, \gate.kr) }, sampleAccurate: true),
		sustain: Pseq([1, 2], inf),
		dur: 3
	)).play
});
)

a.clock.tempo=2

//defaults to false
a.schedSustainInSeconds=true
```

**Comment by thresholdpeople** [comment]
> I noticed that actually it already works as proposed, as the default clock used by Alga is TempoClock.

Ahh yes, you're right! My mistake. That's great then. That was my preferred functionality and I was worried it wouldn't be straightforward to implement. 

And confirming that the variable to switch it to seconds works too. Awesome!

I've also been using some of the things I've been working on and everything seems rock solid so far.

**Comment by vitreo12** [comment]
Glad to hear that!

In any case, I gave the `stretch` / `legato` thing a little more thought and I won't be likely implementing it for 2 reasons:

1. `stretch` already affects `dur` (the `AlgaPattern`'s `Event` derives from the standard one, I am going to remove this dependency at some point).
2. This would introduce yet two more reserved keywords alongside `dur`, `delta`, `sustain`, `fadeTime` and `out`.

What do you think?

**Comment by thresholdpeople** [comment]
So if I understand you correctly, while `stretch` currently works (I personally like that it does), eventually you will remove this functionality so that fewer keywords are tied up in the inner mechanism.

I like `stretch` and more-so `legato`, but really they're conveniences, between `sustain` and `dur`, I do believe it's possible to recreate most of that functionality, and likely have things a bit more readable. 

For me, the biggest feature is not having to set `sustain` explicitly, with it defaulting to `dur` in the case it were not specified. 

I just figured that `legato` would be a nice bonus since `stretch` already seemed to be working.

**Comment by vitreo12** [comment]
>For me, the biggest feature is not having to set sustain explicitly, with it defaulting to dur in the case it were not specified.

That makes sense, I'll look into it

>I just figured that legato would be a nice bonus since stretch already seemed to be working.

Yeah, definitely. I'll see if I can work around it :)

**Comment by vitreo12** [comment]
So, I have added these things:

1. `sustainToDur` flag. This allows to sum the `sustain` to `dur`. So, if `sustain` is 0, the `sustain` that will be used will be `dur`. The default is `false`, in which case that the user needs to provide `sustain` explicitly. If `true`, `sustain` will equal 0 and `dur` will then be used.
2. `stretch` and `legato` now work. Should there be an option to turn off the `stretch` scaling for `sustain`? Or would it be too many options?

What do you think?

**Comment by thresholdpeople** [comment]
I like the "negative" `sustain` capabilities! Might lead to interesting results thinking about it that way. 

But, I think that `sustainToDur` should default to true. My rationale is that if it's true, and someone forgets to set sustain in `AP`, things just "work" and synths on the server don't pile up. If it's set to false by default it's like you have to set it, even if you don't want to set it (eg- the default function of `sustainToDur= true`.. `sustain = dur` is what you want). 

My thought on being able to toggle `stretch` and `legato` on and off is that maybe there's no need? If they don't get set/changed from default, they don't have any effect. Though really I'm ambivalent.

You're looking to avoid a potential namespace clash, right? Then what about picking different, and maybe slightly longer key names for these? Something that's less likely to get used as a variable name to begin with? Maybe something like `durScale` and `eventOverlap`? I only mentioned `stretch` and `legato` because there's precedence in the default `Event`. 

But really if you think it'll cause more harm than good, then maybe it's best to leave out the `stretch` and `legato` functionality all together? That the ability to do `sustain: -0.5` *almost* gets you into the same terrain. Only difference is that the `legato` value sets the relative length, versus sustain is time based. 

And that's the reason I like the idea of `legato` and `stretch`. They allow for a convenient way to scale the overall length, rather than set the duration or hold time. I think musically it's a useful way to think. But it's also not a big deal to add a scaling value to `dur` or `sustain`.

All that being said, I'm not sure `legato` is working correctly for me.
```
(
a = AP((
	def: { SinOsc.ar(\freq.kr(440)) *
		EnvGen.kr(Env.adsr(\atk.kr(0.01), \del.kr(0.3), \sus.kr(0.5), \rel.kr(1.0)), \gate.kr, doneAction: 2) *
		\amp.kr(0.5)
	},
	dur: 4,
	freq: Pseed(1, Pexprand(100.0, 800.0).round(27.3)),
	amp: Pseq([0.3, 0.2], inf),
	// sustain: 4,
	legato: Pseq([1, 0.5], inf),
	rel: 0.1,
	callback: { |ev| ev.postln },
)).play;
a.sustainToDur = true
)
```
With `sustain` commented out, `legato` seems to have no audible effect. With `sustain` in the code, as opposed to releasing the synth about halfway through the first event and holding it for the length of the second, as expected, it seems to add to the overall hold time causing notes to overlap. 

Also, when commenting out `legato`, the `callback` function registers it as being set to 0, but theoretically it should be set to 1, no? Keep `gate` at 1 the entire amount of `dur` or `sustain`.

**Comment by vitreo12** [comment]
> But, I think that sustainToDur should default to true. My rationale is that if it's true, and someone forgets to set sustain in AP, things just "work" and synths on the server don't pile up. If it's set to false by default it's like you have to set it, even if you don't want to set it (eg- the default function of sustainToDur= true.. sustain = dur is what you want).

I totally agree with you, but the problem is that, if the `AlgaSynthDef` already frees itself somehow, you'll get a lot of `FAILURE IN SERVER /n_set Node 1014 not found` errors as the `\gate, 0` sent from `sustain` will be received by a `Node` that's already been freed. Unfortunately, there is no way for me to know how a user would implement its own "self-freeing"capabilities.

Consider this example:

```
(
Alga.boot({
	a = AP((
		def: { SinOsc.ar * EnvPerc.ar(release: 0.5) },
		dur: 1
	)).sustainToDur_(true).play
})
)
```

This is why I'd rather have that as an opt-in feature for users that really need it. Furthermore, if the `AlgaSynthDef` is not using `\gate.kr` to release the `Synth`, it will still keep playing even if the `sustain` is triggered, which might cause confusion.

>With sustain commented out, legato seems to have no audible effect. 

That's my bad, I had a wrong order of operations. It should now work with the newest commit.

```
(
Alga.boot({
	a = AP((
		def: { SinOsc.ar(\freq.kr(440)) *
			EnvGen.kr(Env.adsr(\atk.kr(0.01), \del.kr(0.3), \sus.kr(0.5), \rel.kr(1.0)), \gate.kr, doneAction: 2) *
			\amp.kr(0.5)
		},
		dur: 4,
		freq: Pseed(1, Pexprand(100.0, 800.0).round(27.3)),
		amp: Pseq([0.3, 0.2], inf),
		//sustain: 4,
		legato: Pseq([1, 0.5], inf),
		rel: 0.1,
		callback: { |ev| ev.postln },
	)).sustainToDur_(true).play
})
)

//Same as:
(
Alga.boot({
	a = AP((
		def: { SinOsc.ar(\freq.kr(440)) *
			EnvGen.kr(Env.adsr(\atk.kr(0.01), \del.kr(0.3), \sus.kr(0.5), \rel.kr(1.0)), \gate.kr, doneAction: 2) *
			\amp.kr(0.5)
		},
		dur: 4,
		freq: Pseed(1, Pexprand(100.0, 800.0).round(27.3)),
		amp: Pseq([0.3, 0.2], inf),
		sustain: 4,
		legato: Pseq([1, 0.5], inf),
		rel: 0.1,
		callback: { |ev| ev.postln },
	)).play
})
)
```

**Comment by thresholdpeople** [comment]
> I totally agree with you, but the problem is that, if the AlgaSynthDef already frees itself somehow, you'll get a lot of FAILURE IN SERVER /n_set Node 1014 not found errors as the \gate, 0 sent from sustain will be received by a Node that's already been freed. Unfortunately, there is no way for me to know how a user would implement its own "self-freeing"capabilities.

This makes sense. But could the `sustain` key being explicitly set in the `Event` be a suitable tell? I'm kind of assuming that it will only be there if it needs to be (aside from parameter name overlap). 

Though by keeping the flag:
> 2. stretch and legato now work. Should there be an option to turn off the stretch scaling for sustain? Or would it be too many options?

Perhaps then it can also activate or deactivate `legato`, since `legato` only has a practical effect when the `sustain` parameter is active, maybe that flag can toggle both? And `stretch` can remain active no matter what, or also get flagged, but then it's two flags versus three.

And verifying that the `legato` fix works!

**Comment by vitreo12** [comment]
>This makes sense. But could the sustain key being explicitly set in the Event be a suitable tell? I'm kind of assuming that it will only be there if it needs to be (aside from parameter name overlap).

As of now, if you set `sustain` explicitly, it does work. It simply won't be added to `dur` unless the `sustainToDur` flag is turned on. This flag does not turn on/off the `sustain` mechanism, but it only turns on/off its being added to `dur`.

**Comment by vitreo12** [comment]
By the way, I was thinking that I don't quite like the `sustainToDur` name. Would you have a better suggestion?

**Comment by thresholdpeople** [comment]
I don't have anything right now, but let me think on it!

**Comment by vitreo12** [comment]
Meanwhile, I'll merge this in to develop. We can change the name later before making the release into master.

**Comment by thresholdpeople** [comment]
Sounds good to me!
